File: 2023/D13/part2.py
#!/usr/bin/env python3

#file = open("test", "r")
file = open("input", "r")

def check_horizontal(grid, bug = False):
    for x in range(len(grid) - 1):
        flag = True
        i = x
        j = i + 1
        n = len(grid)
        m = len(grid[0])
        diff = 0
        while i >= 0 and j < n:
            row_1 = grid[i]
            row_2 = grid[j]
            for k in range(m):
                if row_1[k] != row_2[k]:
                    diff += 1
                    flag = False
            i -= 1
            j += 1

        if (bug and diff == 1):
            return x + 1
    return -1

def check_vertical(grid, bug = False):
    for x in range(len(grid[0]) - 1):
        flag = True
        i = x
        j = i + 1
        n = len(grid)
        m = len(grid[0])
        diff = 0
        while i >= 0 and j < m:
            for k in range(n):
                if grid[k][i] != grid[k][j]:
                    diff += 1
                    flag = False
            i -= 1
            j += 1
        if (bug and diff == 1):
            return x + 1
    return -1


grid = [l.strip() for l in file]

patterns = []
pattern = []
for row in grid:
    if row == "":
        patterns.append(pattern)
        pattern = []
    else:
        pattern.append(row)

patterns.append(pattern)
suma = 0
for pattern in patterns:
    hor = check_horizontal(pattern, True)
    ver = check_vertical(pattern, True)
    if hor != -1:
        suma += hor * 100
    if ver != -1:
        suma += ver
print(suma)

File: 2023/D09/part2.hs
#file = open("test", "r")
file = open("input", "r")

def rec(nums, l):
    flag = True
    for num in nums:
        if num != 0:
            flag = False
            break
    if flag:
        return 0
    diff = [0] * (l - 1)
    for i in range(l - 1):
        diff[i] = nums[i + 1] - nums[i]
    ret = rec(diff, l - 1)
    return nums[0] - ret

histories = [list(map(int, l.strip().split(" "))) for l in file]
print(sum([rec(history, len(history)) for history in histories]))
